### PR TITLE
Onboarding new ZetaSQL fuzzers

### PR DESCRIPTION
**Pull request by zd-project**
This PR contains
- one new structure aware fuzzer that fuzzes arithmetic expressions in ZetaSQL.
- fix for UBSAN error.

### CODE REVIEW

**Comment by zd-project** [comment]
Does the above CI build test only build fuzzers or does it also try run them? Looks like the log from libfuzzer, memory, x86_64 caught a memory leak error. Is this intended?

> 2020-07-08T22:47:05.7642950Z Running: docker run --rm --privileged --cap-add SYS_PTRACE -e FUZZING_ENGINE=libfuzzer -e SANITIZER=memory -e ARCHITECTURE=x86_64 -e FUZZING_LANGUAGE=c++ -e MSAN_LIBS_PATH=/work/msan -v /home/runner/work/oss-fuzz/oss-fuzz/build/out/zetasql:/out -v /home/runner/work/oss-fuzz/oss-fuzz/build/work/zetasql:/work -t gcr.io/oss-fuzz/zetasql
2020-07-08T22:47:05.7643689Z Running: docker run --rm --privileged -v /home/runner/work/oss-fuzz/oss-fuzz/build/out/zetasql:/out -v /home/runner/work/oss-fuzz/oss-fuzz/build/work/zetasql:/work -e FUZZING_ENGINE=libfuzzer -e SANITIZER=memory -e ARCHITECTURE=x86_64 -e FUZZING_LANGUAGE=c++ -e MSAN_LIBS_PATH=/work/msan gcr.io/oss-fuzz-base/base-msan-builder patch_build.py /out
> ...
> 2020-07-08T22:49:31.8109373Z DEDUP_TOKEN: malloc--operator new(unsigned long)--operator=
2020-07-08T22:49:31.8110142Z SUMMARY: MemorySanitizer: use-of-uninitialized-value /proc/self/cwd/external/com_google_protobuf/src/google/protobuf/io/tokenizer.cc:123:1 in google::protobuf::io::(anonymous namespace)::Whitespace::InClass(char)
2020-07-08T22:49:31.8110318Z Unique heap origins: 26410
2020-07-08T22:49:31.8110460Z Stack depot allocated bytes: 4749960
2020-07-08T22:49:31.8110597Z Unique origin histories: 16873
2020-07-08T22:49:31.8110737Z History depot allocated bytes: 404952
2020-07-08T22:49:31.8110866Z Exiting
2020-07-08T22:49:31.8111004Z MS: 0 ; base unit: adc83b19e793491b1c6ea0fd8b46cd9f32e592fc

File: projects/zetasql/project.yaml [hunk 1]
```diff
@@ -7,3 +7,4 @@ auto_ccs:
 sanitizers:
```
Comment (evmaus) [inline projects/zetasql/project.yaml:7]: Can we just delete the sanitizer section entirely now that ubsan is working?

File: projects/zetasql/build.sh [hunk 2]
```diff
@@ -49,10 +49,10 @@ fi
 )"
 
 # Temporary hack, see https://github.com/google/oss-fuzz/issues/383
-readonly NO_VPTR='--copt=-fno-sanitize=vptr --linkopt=-fno-sanitize=vptr'
+readonly NO_VPTR='--copt=-fno-sanitize=vptr --linkopt=-fno-sanitize=vptr --cxxopt=-fno-sanitize=vptr'
 
-declare FUZZER_PATH="zetasql/fuzzing/simple_evaluator_fuzzer"
-declare FUZZ_TARGET="//zetasql/fuzzing:simple_evaluator_fuzzer"
+declare FUZZER_PATH=$(for path in $(find zetasql/fuzzing -name *_fuzzer.cc); do echo "${path%.*}"; done)
+declare FUZZ_TARGET="$(for fuzzer in ${FUZZER_PATH}; do echo "//$(dirname ${fuzzer}):$(basename ${fuzzer})"; done)"
```
Comment (evmaus) [inline projects/zetasql/build.sh:55]: So this implies a related naming convention for fuzzer code files & also fuzz targets.  That isn't inherently bad, but it should be documented.  (An alternate option might be to have a naming convention just for the fuzz targets, and use bazel query.)

File: projects/zetasql/Dockerfile [hunk 3]
```diff
@@ -25,6 +25,7 @@ RUN curl https://bazel.build/bazel-release.pub.gpg | apt-key add - && \
 
 RUN apt-get update && apt-get -y install bazel-1.0.0
 
-RUN git clone --depth 1 https://github.com/googleinterns/zetasql-fuzzing.git zetasql
+RUN git clone -b arithmetic_fuzzer --depth 1 \
```
[the remark this answers is not in the document]
Comment (zd-project) [reply projects/zetasql/Dockerfile:28]: I propose that we keep `-b` parameter so we only minimally change the branch name.

File: projects/zetasql/project.yaml (hunk 1 above)
Comment (zd-project) [reply projects/zetasql/project.yaml:7]: Deleting sanitizer disables MSAN as it's not turned on by default.

File: projects/zetasql/build.sh (hunk 2 above)
Comment (zd-project) [reply projects/zetasql/build.sh:55]: We can use `bazel query` instead which is more error-proof.

File: projects/zetasql/project.yaml (hunk 1 above)
Comment (inferno-chromium) [reply projects/zetasql/project.yaml:7]: Yes right. But msan build is failing, can you check the failure.

File: projects/zetasql/build.sh (hunk 2 above)
Comment (inferno-chromium) [reply projects/zetasql/build.sh:55]: Do you plan to do this ? Please ping us when this cl is ready for review and CI passed.

**Comment by zd-project** [comment]
@inferno-chromium @evmaus Since the MSAN is incompatible with LPM as of now (2020/7/14), we'll partially disable MSAN on fuzzers depending on LPM.

Referencing github discussion [here](https://github.com/google/oss-fuzz/issues/864#issuecomment-655203594)

> ...speculation is that MSAN is *only* excluded from LPM but not any underlying library (for example, standard library) it's calling, so even though canary won't be tested in LPM function calls, when it relay data down to other library calls that *are* instrumented with MSAN, error is going to show up again. In our case Bazel may use the standard library that are instrumented with MSAN and that's why MSAN error appears again on invocation of std::min inside LPM

> ...the issue is that LPM has common dependencies with the SUT (e.g., libc++).  Obviously we need to instrument those with MSan, so we need to instrument LPM as well.  I'm thinking the better solution is to unpoison in libfuzzer.

**Comment by inferno-chromium** [comment]
> @inferno-chromium @evmaus Since the MSAN is incompatible with LPM as of now (2020/7/14), we'll partially disable MSAN on fuzzers depending on LPM.
> 
> Referencing github discussion [here](https://github.com/google/oss-fuzz/issues/864#issuecomment-655203594)
> 
> > ...speculation is that MSAN is _only_ excluded from LPM but not any underlying library (for example, standard library) it's calling, so even though canary won't be tested in LPM function calls, when it relay data down to other library calls that _are_ instrumented with MSAN, error is going to show up again. In our case Bazel may use the standard library that are instrumented with MSAN and that's why MSAN error appears again on invocation of std::min inside LPM
> 
> > ...the issue is that LPM has common dependencies with the SUT (e.g., libc++).  Obviously we need to instrument those with MSan, so we need to instrument LPM as well.  I'm thinking the better solution is to unpoison in libfuzzer.

Ok i removed memory from project.yaml, landing this CL. for that bazel query comment, can do it in next cl.

**Pull request merged**